runtime/src/accounts.rs

```diff
@@ -815,48 +793,51 @@ impl Accounts {
         self.accounts_db.add_root(slot)
     }
 
-    fn is_non_loader_key(message: &Message, key: &Pubkey, key_index: usize) -> bool {
+    pub fn is_non_loader_key(message: &Message, key: &Pubkey, key_index: usize) -> bool {
         !message.program_ids().contains(&key) || message.is_key_passed_to_program(key_index)
     }
 
     fn collect_accounts_to_store<'a>(
         &self,
         txs: &'a [Transaction],
         txs_iteration_order: Option<&'a [usize]>,
-        res: &'a [TransactionProcessResult],
-        loaded: &'a mut [(Result<TransactionLoadResult>, Option<HashAgeKind>)],
+        res: &'a [TransactionExecutionResult],
+        loaded: &'a mut [TransactionLoadResult],
         rent_collector: &RentCollector,
         last_blockhash_with_fee_calculator: &(Hash, FeeCalculator),
         fix_recent_blockhashes_sysvar_delay: bool,
         rent_fix_enabled: bool,
     ) -> Vec<(&'a Pubkey, &'a Account)> {
         let mut accounts = Vec::with_capacity(loaded.len());
-        for (i, ((raccs, _hash_age_kind), (_, tx))) in loaded
+        for (i, ((raccs, _nonce_rollback), (_, tx))) in loaded
             .iter_mut()
             .zip(OrderedIterator::new(txs, txs_iteration_order))
             .enumerate()
         {
             if raccs.is_err() {
                 continue;
             }
-            let (res, hash_age_kind) = &res[i];
-            let maybe_nonce = match (res, hash_age_kind) {
-                (Ok(_), Some(HashAgeKind::DurableNonceFull(pubkey, acc, maybe_fee_account))) => {
+            let (res, nonce_rollback) = &res[i];
+            let maybe_nonce = match (res, nonce_rollback) {
+                (Ok(_), Some(nonce_rollback)) => {
+                    let pubkey = nonce_rollback.nonce_address();
+                    let acc = nonce_rollback.nonce_account();
+                    let maybe_fee_account = nonce_rollback.fee_account();
                     Some((pubkey, acc, maybe_fee_account))
```

Comment:
til: it looks like this can be None. That's because we don't use `maybe_nonce` anymore if `res` is `Ok(_)`, considering we enabled `consistent recentblockhashes sysvar`.

Not that I'm asking you to do now with this follow up pr. just a note to solidify my understanding.

Reply:
Yep, using `maybe_nonce` in the `Ok(_)` condition may revert intentional nonce-account state changes.  You do raise a good point though, that this variable is poorly named. I think I'll change it to `maybe_nonce_rollback`